### ISSUE TITLE
Use Poetry for module build and automate PyPi publishing

### ISSUE THREAD

**Issue by george-zubrienko** [lidatong/dataclasses-json#413]
@lidatong what is your opinion on using [Poetry](https://python-poetry.org/)? We use it to keep modules from installing conflicting dependencies (with committed lockfile as well). It will also simplify PyPi publish process.

In addition, as a part of the same change I propose we add a github workflow to publish to PyPi, and put your authorization on it.

Example here https://github.com/SneaksAndData/spark-utils/blob/main/.github/workflows/release.yaml

**Comment by lidatong** [comment]
Yes happy to migrate to Poetry, I've used it myself in newer projects.

> I propose we add a github workflow to publish to PyPi, and put your authorization on it.

This is a good idea, but I need to research a bit the implications of it automatically signing as me. Maybe there's a way to have an "org" do it instead?

**Comment by s-vitaliy** [comment]
@lidatong, @george-zubrienko I was talking about the "Trusted Publishers" PyPI feature: https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/

You can add it to this project using this manual: https://docs.pypi.org/trusted-publishers/adding-a-publisher/

I haven't tried this yet, but it looks very simple to set it up for github actions.

**Comment by george-zubrienko** [comment]
@lidatong ping for the trusted publisher option :) wdyt, should we try?

**Comment by lidatong** [comment]
> @lidatong ping for the trusted publisher option :) wdyt, should we try? For the actual release, we can do something similar to this
> 
> https://github.com/SneaksAndData/github-actions/blob/main/semver_release/semver_release.sh
> 
> Basically we store major/minor in workflow file and bump them manually, patch will be incremented/reset on new minor/major automatically for each github release. That way we can have bugfixes coming from all contributors merged and automatically released on a weekly basis, fully automated, by having a workflow that runs end of every week and checks if there are any commits since last release that could be published.

yep totally makes sense. there's some existing stuff that is actually very similar to what you described (see `publish.sh`, intended to be invoked via `python publish.py 0.5.10`)

i also recently took a pass at creating a github action for releases (see `pythonpackage.yml`)

thanks for proposing!

**Comment by george-zubrienko** [comment]
Amazing, thanks! We'll make a PR for regular scheduled releases next week, will request a review from you :)

**Comment by lidatong** [comment]
> @lidatong, @george-zubrienko I was talking about the "Trusted Publishers" PyPI feature: https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/
> 
> You can add it to this project using this manual: https://docs.pypi.org/trusted-publishers/adding-a-publisher/
> 
> I haven't tried this yet, but it looks very simple to set it up for github actions.
> 
> If you set up a trusted publisher for this package for an environment (let's say 'production') and workflow name (for instance, 'publish.yaml'), I think we can handle the rest.

agreed this seems the way to go, i've been working on setting it up.

one issue with this feature currently is it doesn't work nicely with github action environments (as the security best practice recommends). the problem is the environment is permissioned by branch (following the security best practice) vs. we actually want to release on tagged commit (with the version)

you can see the WIP in `pythonpackage.yml`. i think it can be solved with a separate release workflow file (`release.yml`) with a `on release` trigger, and also setting up the dependency b/w workflows i.e. only run release on test workflow's success

i set up trusted publishing and want to restrict it to run on tag (i.e. intentional releases with a version bump), but it's skipping the pushed tags because they don't match the "environment" and Github

i want it

**Comment by george-zubrienko** [comment]
> i set up trusted publishing and want to restrict it to run on tag (i.e. intentional releases with a version bump), but it's skipping the pushed tags because they don't match the "environment" and Github

You can restrict environment to run only from tagged commits. Thus I think we have everything @s-vitaliy ! We've had a bit of internal mess this week - going to set up releases Monday/Tue and release the newly merged stuff.

**Comment by lidatong** [comment]
> > i set up trusted publishing and want to restrict it to run on tag (i.e. intentional releases with a version bump), but it's skipping the pushed tags because they don't match the "environment" and Github
> 
> You can restrict environment to run only from tagged commits. Thus I think we have everything @s-vitaliy ! We've had a bit of internal mess this week - going to set up releases Monday/Tue and release the newly merged stuff.

This is fixed now! Thanks for the tip, though I didn't see a way in settings to restrict the environment to only run on tagged commit? I made the github action itself check if it's a tagged commit before attempt to run "release". So feel free to push a release by tagging a master commit.

**Comment by george-zubrienko** [comment]
@lidatong Poetry update with with some nice code coverage stats ;)

https://github.com/lidatong/dataclasses-json/pull/431

**Comment by george-zubrienko** [comment]
@lidatong I'm getting this when try to create a GH release - do we need to add some permissions to Actions?

https://github.com/lidatong/dataclasses-json/actions/runs/5567174373/jobs/10168846228

**Comment by lidatong** [comment]
> @lidatong I'm getting this when try to create a GH release - do we need to add some permissions to Actions?
> 
> https://github.com/lidatong/dataclasses-json/actions/runs/5567174373/jobs/10168846228

yes, fixed: https://github.com/lidatong/dataclasses-json/releases/tag/v0.5.10
